design/design-generic-references.md

```diff
@@ -0,0 +1,477 @@
+}
+```
+
+### GenericReference Type
+
+We would like to introduce `GenericReference` field type:
+
+```go
+type GenericReference struct {
+  Reference
+  
+  // ApiVersion of the referenced object.
+  ApiVersion string `json:"apiVersion"`
+  
+  // Kind of the referenced object.
+  Kind string `json:"kind"`
+  
+  // FieldPath of the value within the referenced object.
+  FieldPath string `json:"fieldPath"`
+}
+```
+
+The syntax of the `fieldPath` field is the one used in Composition, in line with
+the [Kubernetes API Conventions on field selection].
+
+Reference fields of the above type describe completely the location of 
+the value that is about to be set on the counterpart field. In case of 
+`datasync.aws.upbound.io` `Task`, an instance might look like:
+
+```yaml
+apiVersion: datasync.aws.upbound.io/v1beta1
+kind: Task
+metadata:
+  name: datasync-task-example
+spec:
+  forProvider:
+    .
+    .
+    sourceLocationArnRef:
+      name: source-location
+      apiVersion: datasync.aws.upbound.io/v1beta1
+      kind: LocationS3
+      fieldPath: metadata.annotations[crossplane.io/external-name]
+```
+
+Such approach lets a user to point the `sourceLocationArnRef` field to a different
+resource at the runtime.
+
+### GenericSelector Type
+
+Similar to `GenericReference`, we are going to introduce `GenericSelector` type:
+
+```go
+type GenericSelector struct {
+  // MatchLabels ensures an object with matching labels is selected.
+  MatchLabels map[string]string `json:"matchLabels,omitempty"`
+  
+  // Policies for selection.
+  // +optional
+  Policy *Policy `json:"policy,omitempty"`
+  
+  // ApiVersion of the referenced object.
+  ApiVersion string `json:"apiVersion"`
+  
+  // Kind of the referenced object.
+  Kind string `json:"kind"`
+  
+  // FieldPath of the value within the referenced object.
+  FieldPath string `json:"fieldPath"`
+}
+```
+
+so that one can reference the external object when its name is
+unknown/irrelevant:
+
+```yaml
+apiVersion: datasync.aws.upbound.io/v1beta1
+kind: Task
+metadata:
+  name: datasync-task-example
+spec:
+  forProvider:
+    .
+    .
+    sourceLocationArnSelector:
+      matchLabels:
+        foo: bar
+      apiVersion: datasync.aws.upbound.io/v1beta1
+      kind: LocationS3
+      fieldPath: metadata.annotations[crossplane.io/external-name]
+```
+
+Although the proposed solution sounds like a natural evolution of the existing
+reference mechanism, its adoption would require that:
+
+* Providers implement the support for it.
+* RBAC permissions need to be relaxed, if provider `A` needs to refer values from
+  instances owned by provider `B`. In case of provider families, each provider has
+  enough rights to access all resources within the family, requiring no changes
+  in RBAC rules. Adding new rules should be done at deployment, by tweaking RBAC
+  rules for the given provider service account.
+
+### Referable Type
+
+If adding a number of RBAC rules is not an option for users, cross-provider
```

Comment:
If I understand this correctly:
- There exists a cross-provider resource reference A->B (a resource in provider A references a resource in provider B)
- Because it's cross-provider, the Crossplane RBAC manager cannot install the rules that will allow A's service account to read B's resources
- The user is not willing to add the RBAC rules herself
- Instead provisions a `Referable` object
- The `Referable` controller is running as a core Crossplane controller and has read access to B's resources: Do we assume here source GVK will be a managed resource?
- `Referable` controller exposes the extracted value in `status` or reports an error if the extraction fails for some reason
- All installed provider's service accounts have read access to all the `Referable` objects in the cluster
- Provider A copies the data from the `Referable` object via a `GenericReference`

So this will allow data transfer across all MRs in the cluster. Do I understand this correctly?